node/pkg/watchers/ibc/watcher.go

```diff
@@ -0,0 +1,700 @@
+	w.channelIdToChainIdLock.Lock()
+	defer w.channelIdToChainIdLock.Unlock()
+	chainID, exists := w.channelIdToChainIdMap[channelID]
+	if exists {
+		return chainID, nil
+	}
+
+	// We continue to hold the lock here because we don't want two routines (event handler and reobservation) both querying at the same time.
+	channelIdToChainIdMap, err := w.queryChannelIdToChainIdMapping()
+	if err != nil {
+		w.logger.Error("failed to query channelID to chainID mapping", zap.Error(err))
+		return vaa.ChainIDUnset, err
+	}
+
+	w.channelIdToChainIdMap = channelIdToChainIdMap
+
+	chainID, exists = w.channelIdToChainIdMap[channelID]
+	if exists {
+		return chainID, nil
+	}
+
+	return vaa.ChainIDUnset, nil
+}
+
+/*
+This query:
+`"all_channel_chains"` is `ImFsbF9jaGFubmVsX2NoYWlucyI=`
+which becomes:
+http://localhost:1319/cosmwasm/wasm/v1/contract/wormhole1nc5tatafv6eyq7llkr2gv50ff9e22mnf70qgjlv737ktmt4eswrq0kdhcj/smart/ImFsbF9jaGFubmVsX2NoYWlucyI%3D
+
+Returns something like this:
+{
+  "data": {
+    "channels_chains": [
+      [
+        "Y2hhbm5lbC0w",
+        18
+      ]
+    ]
+  }
+}
+
+*/
+
+type ibcAllChannelChainsQueryResults struct {
+	Data struct {
+		ChannelChains [][]interface{} `json:"channels_chains"`
+	}
+}
+
+var allChannelChainsQuery = url.QueryEscape(base64.StdEncoding.EncodeToString([]byte(`"all_channel_chains"`)))
+
+// queryChannelIdToChainIdMapping queries the contract for the set of IBC channels and their correspond chain IDs.
+func (w *Watcher) queryChannelIdToChainIdMapping() (map[string]vaa.ChainID, error) {
+	client := &http.Client{
+		Timeout: time.Second * 5,
+	}
+
+	query := fmt.Sprintf(`%s/cosmwasm/wasm/v1/contract/%s/smart/%s`, w.lcdUrl, w.contractAddress, allChannelChainsQuery)
+	resp, err := client.Get(query)
+	if err != nil {
+		return nil, fmt.Errorf("query failed: %w", err)
+	}
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("read failed: %w", err)
+	}
+	resp.Body.Close()
+
+	var result ibcAllChannelChainsQueryResults
+	err = json.Unmarshal(body, &result)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal response: %s, error: %w", string(body), err)
+	}
+
+	if len(result.Data.ChannelChains) == 0 {
+		return nil, fmt.Errorf("query did not return any data")
+	}
+
+	w.logger.Info("queried IBC channel ID mapping", zap.Int("numEntriesReturned", len(result.Data.ChannelChains)))
+
+	ret := make(map[string]vaa.ChainID)
+	for idx, entry := range result.Data.ChannelChains {
+		if len(entry) != 2 {
+			return nil, fmt.Errorf("channel map entry %d contains %d items when it should contain exactly two, json: %s", idx, len(entry), string(body))
+		}
+
+		channelIdBytes, err := base64.StdEncoding.DecodeString(entry[0].(string))
+		if err != nil {
+			return nil, fmt.Errorf("channel ID for entry %d is invalid base64: %s, err: %s", idx, entry[0], err)
+		}
+
+		channelID := string(channelIdBytes)
+		chainID := vaa.ChainID(entry[1].(float64))
```

Comment:
Why do we parse as `float64` here instead of `uint16`?

Reply:
That is the data type that is returned. It won't compile without float64.